kratos/geometries/nurbs_curve_geometry.h

```diff
@@ -0,0 +1,488 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Thomas Oberbichler
+//
+//  Ported from the ANurbs library (https://github.com/oberbichler/ANurbs)
+//
+
+#if !defined(KRATOS_NURBS_CURVE_GEOMETRY_H_INCLUDED )
+#define  KRATOS_NURBS_CURVE_GEOMETRY_H_INCLUDED
+
+// Project includes
+#include "geometries/geometry.h"
+
+#include "geometries/nurbs_shape_function_utilities/nurbs_curve_shape_functions.h"
+#include "geometries/nurbs_shape_function_utilities/nurbs_interval.h"
+
+
+namespace Kratos {
+
+template <int TWorkingSpaceDimension, class TContainerPointType>
+class NurbsCurveGeometry : public Geometry<typename TContainerPointType::value_type>
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Geometry as base class.
+    typedef Geometry<typename TContainerPointType::value_type> BaseType;
+    typedef NurbsCurveGeometry<TWorkingSpaceDimension, TContainerPointType> GeometryType;
+
+    typedef typename BaseType::IndexType IndexType;
+    typedef typename BaseType::SizeType SizeType;
+
+    /** Array of counted pointers to point. This type used to hold
+        geometry's points.*/
+    typedef  typename BaseType::PointsArrayType PointsArrayType;
+
+    typedef  typename BaseType::CoordinatesArrayType CoordinatesArrayType;
+
+    /// Counted pointer of NurbsCurveGeometry
+    KRATOS_CLASS_POINTER_DEFINITION(NurbsCurveGeometry);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Conctructor for B-Spline curves
+    NurbsCurveGeometry(
+        const PointsArrayType& rThisPoints,
+        const SizeType PolynomialDegree,
+        const Vector& rKnots)
+        : BaseType(rThisPoints, &msGeometryData)
+        , mPolynomialDegree(PolynomialDegree)
+        , mKnots(rKnots)
+    {
+        KRATOS_DEBUG_ERROR_IF(rKnots.size() != NurbsUtilities::GetNumberOfKnots(PolynomialDegree, rThisPoints.size()))
```

Comment:
maybe better make this a release error since it is a significant error/misuse
depends on how often this is being called

same for the other constructors